File: frontend/src/features/practice/runner-machine.ts
import type { QuestionType, PreviousAnswer, AnswerResult } from "@/lib/api/types";

export type RunnerPhase = "selecting" | "submitted";

export interface RunnerState {
  phase: RunnerPhase;
  selected: number[];
  result: AnswerResult | null;
}

export function initialRunnerState(previous?: PreviousAnswer | null): RunnerState {
  if (previous) {
    return { phase: "submitted", selected: [...previous.selected], result: null };
  }
  return { phase: "selecting", selected: [], result: null };
}

export function toggleSelection(
  state: RunnerState,
  orderIndex: number,
  questionType: QuestionType
): RunnerState {
  if (state.phase !== "selecting") return state;
  if (questionType === "multiple_choice") {
    const has = state.selected.includes(orderIndex);
    const selected = has
      ? state.selected.filter((i) => i !== orderIndex)
      : [...state.selected, orderIndex].sort((a, b) => a - b);
    return { ...state, selected };
  }
  // single_choice, true_false, and any single-answer type: replace
  return { ...state, selected: [orderIndex] };
}

export function canSubmit(state: RunnerState): boolean {
  return state.phase === "selecting" && state.selected.length > 0;
}

export function markSubmitted(state: RunnerState, result: AnswerResult): RunnerState {
  return { phase: "submitted", selected: state.selected, result };
}

File: mobile/lib/features/practice/runner_machine.dart
import 'package:cissp_api/cissp_api.dart';

enum RunnerPhase { selecting, submitted }

class RunnerState {
  const RunnerState({
    required this.phase,
    required this.selected,
    this.result,
  });

  final RunnerPhase phase;
  final List<int> selected;
  final AnswerResult? result;
}

RunnerState initialRunnerState(PreviousAnswer? previous) {
  if (previous != null) {
    return RunnerState(
      phase: RunnerPhase.submitted,
      selected: List<int>.from(previous.selected),
    );
  }
  return const RunnerState(phase: RunnerPhase.selecting, selected: []);
}

RunnerState toggleSelection(
  RunnerState state,
  int orderIndex,
  String questionType,
) {
  if (state.phase != RunnerPhase.selecting) return state;
  if (questionType == 'multiple_choice') {
    final has = state.selected.contains(orderIndex);
    final selected = has
        ? state.selected.where((i) => i != orderIndex).toList()
        : [...state.selected, orderIndex]..sort();
    return RunnerState(phase: state.phase, selected: selected, result: state.result);
  }
  return RunnerState(
    phase: state.phase,
    selected: [orderIndex],
    result: state.result,
  );
}

bool canSubmit(RunnerState state) =>
    state.phase == RunnerPhase.selecting && state.selected.isNotEmpty;

RunnerState markSubmitted(RunnerState state, AnswerResult result) {
  return RunnerState(
    phase: RunnerPhase.submitted,
    selected: state.selected,
    result: result,
  );
}
